src/execution_plan/record_map.c

```diff
@@ -0,0 +1,209 @@
+    if (type == CYPHER_AST_IDENTIFIER) {
+        alias = cypher_ast_identifier_get_name(entity);
+    } else if (type == CYPHER_AST_NODE_PATTERN) {
+        const cypher_astnode_t *ast_alias = cypher_ast_node_pattern_get_identifier(entity);
+        if (ast_alias) alias = cypher_ast_identifier_get_name(ast_alias);
+    } else if (type == CYPHER_AST_REL_PATTERN) {
+        const cypher_astnode_t *ast_alias = cypher_ast_rel_pattern_get_identifier(entity);
+        if (ast_alias) alias = cypher_ast_identifier_get_name(ast_alias);
+    } else if (type == CYPHER_AST_PROPERTY_OPERATOR) {
+        const cypher_astnode_t *ast_expr = cypher_ast_property_operator_get_expression(entity);
+        if (cypher_astnode_type(ast_expr) == CYPHER_AST_IDENTIFIER) {
+            alias = cypher_ast_identifier_get_name(ast_expr);
+        }
+    } else {
+        assert(false);
+    }
+    if (alias) {
+        id_ptr = _BuildMapValue(id);
+        TrieMap_Add(record_map->map, (char*)alias, strlen(alias), id_ptr, TrieMap_DONT_CARE_REPLACE);
+        // printf("Alias %s at %d from FindOrAddASTEntity\n", alias, id);
+    }
+    // Map AR_ExpNode?
+
+    return id;
+}
+
+uint RecordMap_FindOrAddID(RecordMap *record_map, uint entity_id) {
+    // Ensure this is a new entity
+    // assert(TrieMap_Find(record_map->map, (char*)&entity, sizeof(entity)) == TRIEMAP_NOTFOUND);
+
+    uint *id_ptr = TrieMap_Find(record_map->map, (char*)&entity_id, sizeof(entity_id));
+    if (id_ptr != TRIEMAP_NOTFOUND) return *id_ptr;
+
+    uint id = record_map->record_len++;
+
+    // Map ID value
+    id_ptr = _BuildMapValue(id);
+    TrieMap_Add(record_map->map, (char*)&entity_id, sizeof(entity_id), id_ptr, TrieMap_DONT_CARE_REPLACE);
+
+    return id;
+}
+
+uint RecordMap_FindOrAddAlias(RecordMap *record_map, const char *alias) {
+    // This alias may already be represented in the Record map
+    uint *id_ptr = TrieMap_Find(record_map->map, (char*)alias, strlen(alias));
+    if (id_ptr != TRIEMAP_NOTFOUND) return *id_ptr;
+
+    // TODO tmp
+    AST *ast = AST_GetFromTLS();
+    uint record_id;
+
+    // Retrieve the AST ID
+    uint ast_id = AST_GetEntityIDFromAlias(ast, alias);
+    // The AST ID will not exist if this alias is projected by a WITH clause
+    if (ast_id != IDENTIFIER_NOT_FOUND) {
+        // This AST ID may already be represented in the record map
+        record_id = RecordMap_LookupEntityID(record_map, ast_id);
+        if (record_id != IDENTIFIER_NOT_FOUND) return record_id;
+    }
+
+    // if (record_id != IDENTIFIER_NOT_FOUND) {
+        // id_ptr = _BuildMapValue(record_id);
+        // TrieMap_Add(record_map->map, alias, strlen(alias), id_ptr, TrieMap_DONT_CARE_REPLACE);
+        // return *id_ptr;
+    // }
+
+    // Generate new Record ID
+    uint id = record_map->record_len++;
+
+    // Map alias to Record ID
+    id_ptr = _BuildMapValue(id);
+    TrieMap_Add(record_map->map, (char*)alias, strlen(alias), id_ptr, TrieMap_DONT_CARE_REPLACE);
```

Comment:
I need to think about this. I don't believe it's possible to get into an inconsistency in this region, but there are lots of scenarios where assumptions about a value existing (justifying an `asssert`) are invalid. For example, a WITH projection has an alias, but not an AST ID.

I've changed the values to be casted integers rather than heap allocations, so no leaks are possible here.